Code/Tools/ProjectManager/Source/FormLineEditTagsWidget.cpp

```diff
@@ -0,0 +1,201 @@
+
+        //add the drop down button for completion options
+        m_dropdownButton = new QPushButton(QIcon(":/CarrotArrowDown.svg"), "", this);
+        m_dropdownButton->setObjectName("dropDownButton");
+        m_frameLayout->addWidget(m_dropdownButton);
+        connect(m_dropdownButton, &QPushButton::clicked, this, [=]([[maybe_unused]]bool ignore){ m_lineEdit->completer()->complete(QRect()); });
+
+        //section of form for showing tags
+        m_tagFrame = new QFrame(this);
+        m_tagFrame->setObjectName("formTagField");
+        
+        QHBoxLayout* tagsLayout = new QHBoxLayout();
+        tagsLayout->setSpacing(8);
+        tagsLayout->addStretch();
+        
+        m_tagFrame->setLayout(tagsLayout);
+        m_tagFrame->setVisible(false);
+        tagsLayout->setSizeConstraint(QLayout::SetNoConstraint);
+
+        m_mainLayout->addWidget(m_tagFrame);
+    }
+
+    FormLineEditTagsWidget::FormLineEditTagsWidget(const QString& labelText, const QString& valueText, QWidget* parent)
+        : FormLineEditTagsWidget(labelText, valueText, "", "", parent)
+    {
+    }
+
+    void FormLineEditTagsWidget::forceSetText(const QString& text)
+    {
+        m_lineEdit->setText(text);
+        m_lineEdit->setFocus();
+    }
+
+    void FormLineEditTagsWidget::forceSubmitCurrentText()
+    {
+        addToTagList(m_lineEdit->text());
+        m_lineEdit->clear();
+        m_lineEdit->setText("");
+        refreshTagFrame();
+    }
+
+    void FormLineEditTagsWidget::refreshTagFrame()
+    {
+        // cleanup the tag frame widget and re-add the tag list
+        qDeleteAll(m_tagFrame->children());
+        QHBoxLayout* layout = new QHBoxLayout(this);
+        layout->setSpacing(8);
+
+        for (auto t : m_tags)
+        {
+            QCheckBox* tc = new QCheckBox(t, this);
+            tc->setLayoutDirection(Qt::RightToLeft);
+            // connect the checked signal to a good slot
+            connect(tc, &QCheckBox::stateChanged, this, &FormLineEditTagsWidget::processTagDelete);
+            layout->addWidget(tc);
+        }
+
+        layout->addStretch();
+        layout->setSizeConstraint(QLayout::SetNoConstraint);
+        m_tagFrame->setLayout(layout);
+
+        bool makeVisible = m_tags.count() > 0;
+        m_tagFrame->setVisible(makeVisible);
+        refreshStyle();
+    }
+
+    void FormLineEditTagsWidget::addToTagList(const QString& text)
+    {
+        if((QString::compare(text,"") > 0 || QString::compare(text,"") < 0) && !m_tags.contains(text))
+        {
+            m_tags << text;
+        }
+    }
+
+    void FormLineEditTagsWidget::textChanged(const QString& text)
+    {
+        if(text.contains(" "))
+        {
+            //do tag adding
+            QString interim = m_lineEdit->text();
+            m_lineEdit->clear();
+            for(auto str : interim.split(" "))
+            {
+                addToTagList(str);
+            }
+            refreshTagFrame();
+        }
+    } 
```

Comment:
So when the interim string is split by space, there are 3 distinct cases that the text line will detect ( _ as space):

1.  _word

2.  word_

3.  wo_rd

In the first case, the validator regex actually prevents a space from ever preceding before the word. The line edit must always start with a non-whitespace character.

In the second cases the "word" is picked up as an entire string tag, while the latter part is basically an empty string, which gets filtered out in `addToTagList`. 

In the third case "wo" and "rd" are two separate strings, and rendered as tags separately. This was verified in testing the tag field.

All cases get the intended text pieces as tags.

Reply:
I was watching the gif in this PR, and it shows "AtomRendering" being written, then a space being added in the middle, and only "Atom" becoming a tag. Rewatching, I realized "Rendering" was already a tag, so I guess that's why it wasn't added again - and what caused my confusion. 

I wonder if we could clarify the situation when users try to add an already existing tag... but that's more of a follow-up question than anything.

Reply:
Agreed, that is a follow up item, because primary correctness is not affected in this PR. Most likely I would need to add an additional 'information frame' that would allow me to write messages such as :
"Note: Rendering tag already exists"

And then it fades after 3 seconds. I would most likely add it to the bottom of the tags.